posts/tradeoff-of-defining-types-as-subobjects.md

```diff
@@ -0,0 +1,119 @@
+---
+author: 'Yaël Dillies'
+category: 'Papers'
+date: 2024-07-25 14:00:00 UTC+03:00
+description: ''
+has_math: true
+link: ''
+slug: tradeoff-of-defining-types-as-subobjects
+tags: ''
+title: Tradeoffs of defining types as subobjects
+type: text
+---
+
+It often happens in formalisation that a type of interest is a subobject of another type of interest. For example, the unit circle in the complex plane is naturally a submonoid (nevermind that it is actually a subgroup, we currently can't talk about subgroups of fields).
+
+What is the best way of defining this unit circle in terms of `Complex`? This blog post examines the pros and cons of the available designs.
+
+## The available designs
+
+We will illustrate the various designs using the "circle in the complex plane" example.
+
+The first option, most loyal to the description of the unit circle as a submonoid of the complex plane, is to simply define the unit circle as a `Submonoid Complex`. We will call the "Subobject design".
+
+```lean
+-- Subobject design
+def circle : Submonoid Complex where
+  carrier := {x : Complex | ‖x‖ = 1}
+  one_mem' := sorry
+  mul_mem' := sorry
+```
+
+The second obvious choice would be to define the unit circle as a custom structure. We call this the "Custom Structure" design.
+
+```lean
+-- Custom Structure design
+structure Circle : Type where
+  val : Complex
+  norm_val : ‖val‖ = 1
+```
+
+Finally, an intermediate option would be to define it as the coercion to `Type` of a subobject. We call this the "Coerced Subobject" design.
+
+```lean
+-- Coerced Subobject design
+def Circle : Type := circle -- possibly replacing `circle` by its definition
+```
+
+## Typeclass search
+
+In the Subobject design, all instances about `circle` are actually about `↥circle` where `↥` is [`CoeSort.coe`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=CoeSort.coe#doc). Eg `Foo circle` is secretly `Foo ↥circle`. This makes the head symbol be `CoeSort.coe`, meaning that typeclass search will try unifying *every* `Foo ↥someSubobject` instance with `Foo ↥circle`. In doing so, it will try unifying *every* subobject `someSubobject` with a `Foo` instance with `circle`. This is potentially really expensive.
```

Comment:
Is this still the case in Lean 4? I thought discrimination trees avoided this...

Reply:
Very much so. Else why would https://github.com/leanprover-community/mathlib4/pull/12386 have achieved anything perf-wise? The problem I believe is that the discr tree key is `[CoeSort.coe, *]`